### PR TITLE
Enable calibration, callbacks and custom commands for EZO sensors

### PR DESCRIPTION
**Pull request by gvdhoven**
# What does this implement/fix?

There are some long delays in the sensors API and callbacks work great for this. The esphome ezo API has been expanded to support a few more commands/callbacks as well as being able to send any custom command to the EZO boards. The below example entry yaml shows what the callbacks look like. **There are no breaking changes, code has been made backwards-compatible.**

_**Note: I am not the original author; that is SenexCrenshaw, however I had to remove the que-ing from the PR in order to get it working and refactored some code. It has been tested on an ESP32 with an pH probe and ORP probe.**_

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes n/a

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2133

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:
```yaml
  - platform: ezo
    id: ph_ezo
    address: 99
    unit_of_measurement: "pH"
    update_interval: 60s
    state_topic: suites/pool/esp/ezo/state
    on_led:
      - lambda: |-
          id(mqtt_client).publish("suites/pool/esp/ezo/led/state", x ? "ON" : "OFF");
    on_device_information:
      - lambda: |-
          id(mqtt_client).publish("suites/pool/esp/ezo/deviceinformation/state", x );
    on_slope:
      - lambda: |-
          id(mqtt_client).publish("suites/pool/esp/ezo/slope/state", x );
    on_calibration:
      - lambda: |-
          id(mqtt_client).publish("suites/pool/esp/ezo/calibration/state", x );
    on_t:
      - lambda: |-
          id(mqtt_client).publish("suites/pool/esp/ezo/t/state", x );
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by gvdhoven** [comment]
@PoppyPop can you confirm this new code works? I don't have a working setup at the moment to test this.

@ssieb asked me via Discord:

> Please test out the updated code.  Try changing the i2c address and make sure it continues working.  Test it with a small update interval over time and make sure there's no memory leak.  Add the https://esphome.io/components/debug.html component and watch the free sensor.

File: esphome/components/ezo/ezo.h [hunk 1]
```diff
@@ -15,13 +40,71 @@ class EZOSensor : public sensor::Sensor, public PollingComponent, public i2c::I2
   void update() override;
   float get_setup_priority() const override { return setup_priority::DATA; };
 
-  void set_tempcomp_value(float temp);
+  // I2C
+  void set_address(unsigned int address);
```
Comment (jesserockz) [inline esphome/components/ezo/ezo.h]:
```suggestion
  void set_address(uint8_t address);
```

File: esphome/components/ezo/ezo.cpp [hunk 2]
```diff
@@ -66,28 +104,142 @@ void EZOSensor::loop() {
...
+  ezo_command->delay_ms = delay_ms;
+  this->commands_.push_back(std::move(ezo_command));
+};
+
+void EZOSensor::set_calibration_point_(EzoCalibrationType type, float value) {
+  std::string payload = str_sprintf("Cal,%s,%0.2f", EZO_CALIBRATION_TYPE_STRINGS[type], value);
+  this->add_command_(payload, EzoCommandType::EZO_CALIBRATION, 900);
+}
+
+void EZOSensor::set_address(unsigned int address) {
```
Comment (jesserockz) [inline esphome/components/ezo/ezo.cpp]:
```suggestion
void EZOSensor::set_address(uint8_t address) {
```

**Review by jesserockz — APPROVED** [review]
Looks good, minor nitpick =)

**Comment by PoppyPop** [comment]
I was on holiday, i will check this tonight (GMT+2), and report the free value tomorrow. 
The system was working rock solid for me since 2 months with an update interva of 60sec.

I make the test with an update interval of ~~1s~~ 2s. (At 1sec, the sensor don't have the time to respond)

For memory, the value at startup
Heap Free | 24480 B |  
Heap Max Block | 20592 B

**Comment by PoppyPop** [comment]
16 hours after : Heap Free | 24488 B

No reboot, no problem

**Pull request merged**